gym_roboy/envs/tests/test_msj_env.py

```diff
@@ -101,3 +103,44 @@ def test_msj_env_reward_is_lower_with_joint_vel_penalty():
 
 def test_msj_env_render_does_nothing(msj_env):
     msj_env.render()
+
+
+@pytest.mark.parametrize(
+    "env",
+    [MsjEnv(ros_proxy=MockMsjROSProxy(), joint_vel_penalty=True),
+     MsjEnv(ros_proxy=MockMsjROSProxy(), joint_vel_penalty=False)],
+    ids=["with joint_vel penalty", "no joint_vel penalty"]
+)
+def test_msj_env_reward_monotonously_improves_during_approach(env: MsjEnv):
+    """
+    For any random starting state, if we approach the goal at every
+    step, the reward should strictly improve. This is like computing
+    the numerical gradient dr/dq and checking that it is positive as
+    the state approaches the goal state.
+
+    This tests catches mistakes like reaching optimality by not moving.
+    """
+    np.random.seed(0)
+    env.seed(0)
+
+    num_starting_states = 40  # roughly covers many points in the space
+    num_steps = 7  # distance to goal halves at every step
+
+    goal_state = MsjRobotState.new_random_zero_vel_state()
+
+    starting_states = [MsjRobotState.new_random_state() for _ in range(num_starting_states)]
+    starting_states.append(MsjRobotState.new_random_zero_vel_state())
+    if env._joint_vel_penalty:
+        starting_states.append(MsjRobotState.new_random_zero_angle_state())
```

Comment:
I dont understand this. If we have joint vel penalty dont we need zero vel state to check if it is reaching optimum with zero velocity?

Reply:
@BarisYazici 
A `new_random_zero_vel_state` has perfect `joint_vels` (0), but improvable `joint_angles` (!=0), so we must always check that reward improves as it approaches the `goal_state` (with or without `_joint_vel_penalty`, line 132) 

In contrast to that, a `new_random_zero_angle_state` has perfect `joint_angles` (0) and there is no room for improvement. That is unless we have set `_joint_vel_penalty=True`, in that case, the `new_random_zero_angle_state` has improvable `joint_vels` (!=0), so we must check that reward improves as it approaches the `goal_state` (lines 133, 134).